File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.go
func atMost(nums []int, goal int) int {
    if goal < 0 {
        return 0
    }
    cnt,sum, l := 0, 0, 0

    for r:=0; r < len(nums); r++ {
        sum += nums[r]
        for sum > goal {
            sum -= nums[l]
            l += 1
        }
        cnt += r - l + 1
    }
    return cnt
}

func numSubarraysWithSum(nums []int, goal int) int {
    return atMost(nums, goal) - atMost(nums, goal - 1)
}

File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.rs
impl Solution {
    pub fn num_subarrays_with_sum(nums: Vec<i32>, goal: i32) -> i32 {
        Self::atmost(&nums, goal) - Self::atmost(&nums, goal - 1)
    }

    pub fn atmost(nums: &Vec<i32>, goal: i32) -> i32 {
        if goal < 0 {
            return 0;
        }
        let (mut cnt, mut sum, mut l) = (0, 0, 0);

        for r in 0..nums.len() {
            sum += nums[r];
            while sum > goal {
                sum -= nums[l];
                l += 1;
            }
            cnt += r - l + 1;
        }
        cnt as i32
    }
}
